buildroot/share/PlatformIO/scripts/preflight-checks.py

```diff
@@ -0,0 +1,62 @@
+#
+# preflight-checks.py
+# Script to check for common issues prior to compiling
+#
+import os
+import re
+Import("env")
+
+def get_envs_for_board(board):
+    if board.startswith("BOARD_"):
+        board = board[6:]
+    with open("Marlin/src/pins/pins.h","r") as f:
+        board_found = ""
+        r=re.compile(r"if\s+MB\((.+)\)")
+        for line in f.readlines():
+            mbs = r.findall(line)
+            if mbs:
+                board_found = board if board in re.split(r",\s*", mbs[0]) else ""
+            if board_found and "#include " in line and "env:" in line:
+                return re.findall(r"env:\w+", line)
+    return []
+
+def check_envs(build_env, base_envs, config):
+    if build_env in base_envs:
+        return True
+    ext = config.get(build_env, 'extends', default=None)
+    if ext:
+        for ext_env in ext:
+            if check_envs(ext_env, base_envs, config):
+                return True
+    return False
+
+# Sanity checks:
+if 'PIOENV' not in env:
+    raise SystemExit("Error: PIOENV is not defined. This script is intended to be used with PlatformIO")
+
+if 'MARLIN_FEATURES' not in env:
+    raise SystemExit("Error: this script should be used after common Marlin scripts")
+
+if 'MOTHERBOARD' not in env['MARLIN_FEATURES']:
+    raise SystemExit("Error: MOTHERBOARD is not defined in Configuration.h")
+
+build_env = env['PIOENV']
+motherboard = env['MARLIN_FEATURES']['MOTHERBOARD']
+base_envs = get_envs_for_board(motherboard)
+config = env.GetProjectConfig()
+result = check_envs("env:"+build_env, base_envs, config)
+
+if not result:
+    err = "Error: your selected build environment '%s' is not compatible with MOTHERBOARD=%s in Configuration.h. " \
+          "Please use one of compatible build environments for this board: %s" % \
+          (build_env, motherboard, ",".join([e[4:] for e in base_envs if e.startswith("env:")]))
+    raise SystemExit(err)
+
+#
+# Check for Config files in two common incorrect places
+#
+for p in [ env['PROJECT_DIR'], os.path.join(env['PROJECT_DIR'], "config") ]:
+	for f in [ "Configuration.h", "Configuration_adv.h" ]:
+		if os.path.isfile(os.path.join(p, f)):
+			err = 'ERROR: Config files found in directory ' + str(p) + '. Please move them into the Marlin subfolder.'
```

Comment:
please, don't use `str(p)`. `p` is already a string. And for nice formatting it is better this way:

```suggestion
			err = "ERROR: Config files found in directory %s. Please move them into the Marlin subfolder." % p
```

Reply:
The internet says you shouldn't be using % either but .format..  but I'm not a python expert (clearly)

Reply:
It depends on the willingness to maintain compatibility with older Pythons in complex cases. The .format() is indeed new way, but it was initially introduced in python3, then backported to 2.6/2.7 and over course of 3.x releases .format() syntax evolved.
For such small cases like above, inserting single string, it wouldn't make much difference. But yes, going forward it might be good to convert to Python3 conventions.